File: sdk/javascript/src/merkle.js
/**
 * FoodBlock Merkle — Merkle-ized state for selective disclosure.
 * Each state field becomes a leaf in a Merkle tree.
 */

const crypto = require('crypto')

/**
 * Compute SHA-256 hash of a string.
 * @param {string} data
 * @returns {string} hex digest
 */
function sha256(data) {
  return crypto.createHash('sha256').update(data, 'utf8').digest('hex')
}

/**
 * Produce a canonical string representation of a value for hashing.
 * @param {*} value
 * @returns {string}
 */
function canonicalValue(value) {
  if (value === null || value === undefined) return 'null'
  if (typeof value === 'object') return JSON.stringify(value, Object.keys(value).sort())
  return String(value)
}

/**
 * Create a Merkle tree from a state object.
 * Each key-value pair becomes a leaf: SHA-256(key + ":" + canonical(value)).
 *
 * @param {object} state - The state object to merkle-ize
 * @returns {object} { root: hex, leaves: { fieldName: leafHash }, tree: [layer0, layer1, ...] }
 */
function merkleize(state) {
  if (!state || typeof state !== 'object') {
    throw new Error('FoodBlock: state must be an object')
  }

  const keys = Object.keys(state).sort()
  const leaves = {}

  for (const key of keys) {
    leaves[key] = sha256(key + ':' + canonicalValue(state[key]))
  }

  // Build tree layers
  const layer0 = keys.map(k => leaves[k])
  const tree = [layer0]

  let currentLayer = layer0
  while (currentLayer.length > 1) {
    const nextLayer = []
    for (let i = 0; i < currentLayer.length; i += 2) {
      if (i + 1 < currentLayer.length) {
        // Pair two nodes — sort for deterministic ordering
        const pair = [currentLayer[i], currentLayer[i + 1]].sort()
        nextLayer.push(sha256(pair[0] + pair[1]))
      } else {
        // Odd node — promote to next layer
        nextLayer.push(currentLayer[i])
      }
    }
    tree.push(nextLayer)
    currentLayer = nextLayer
  }

  const root = currentLayer.length > 0 ? currentLayer[0] : sha256('')

  return { root, leaves, tree }
}

/**
 * Create a selective disclosure of specific fields with a Merkle proof.
 *
 * @param {object} state - The full state object
 * @param {string[]} fieldNames - Fields to disclose
 * @returns {object} { disclosed, proof, root }
 */
function selectiveDisclose(state, fieldNames) {
  if (!state || typeof state !== 'object') {
    throw new Error('FoodBlock: state must be an object')
  }
  if (!Array.isArray(fieldNames)) {
    throw new Error('FoodBlock: fieldNames must be an array')
  }

  const { root, leaves, tree } = merkleize(state)

  const disclosed = {}
  for (const name of fieldNames) {
    if (name in state) {
      disclosed[name] = state[name]
    }
  }

  // Collect proof: sibling hashes needed to reconstruct the root
  const sortedKeys = Object.keys(state).sort()
  const proof = []

  for (const name of fieldNames) {
    const idx = sortedKeys.indexOf(name)
    if (idx === -1) continue

    // Walk up the tree collecting sibling nodes
    let currentIdx = idx
    for (let layer = 0; layer < tree.length - 1; layer++) {
      const layerNodes = tree[layer]
      const siblingIdx = (currentIdx % 2 === 0) ? currentIdx + 1 : currentIdx - 1

      if (siblingIdx >= 0 && siblingIdx < layerNodes.length) {
        proof.push({
          hash: layerNodes[siblingIdx],
          position: currentIdx % 2 === 0 ? 'right' : 'left',
          layer
        })
      }
      // Move to parent index
      currentIdx = Math.floor(currentIdx / 2)
    }
  }

  return { disclosed, proof, root }
}

/**
 * Verify that disclosed fields and proof reconstruct the given Merkle root.
 *
 * @param {object} disclosed - Object with disclosed field key-value pairs
 * @param {object[]} proof - Array of { hash, position, layer } proof elements
 * @param {string} root - Expected Merkle root
 * @returns {boolean} True if the proof is valid
 */
function verifyProof(disclosed, proof, root) {
  if (!disclosed || typeof disclosed !== 'object') return false
  if (!root || typeof root !== 'string') return false

  // Recompute leaf hashes for each disclosed field
  const disclosedKeys = Object.keys(disclosed).sort()

  for (const key of disclosedKeys) {
    let currentHash = sha256(key + ':' + canonicalValue(disclosed[key]))

    // Gather proof entries in layer order for this field
    const relevantProofs = proof.filter(() => true) // use all proof elements
    const byLayer = {}
    for (const p of relevantProofs) {
      if (!byLayer[p.layer]) byLayer[p.layer] = []
      byLayer[p.layer].push(p)
    }

    // Walk up layers
    const maxLayer = Math.max(...proof.map(p => p.layer), -1)
    for (let layer = 0; layer <= maxLayer; layer++) {
      const layerProofs = byLayer[layer]
      if (!layerProofs || layerProofs.length === 0) continue

      // Find the proof element for this layer
      const proofEntry = layerProofs.shift()
      if (!proofEntry) continue

      const pair = proofEntry.position === 'right'
        ? [currentHash, proofEntry.hash]
        : [proofEntry.hash, currentHash]

      const sorted = pair.sort()
      currentHash = sha256(sorted[0] + sorted[1])
    }

    if (currentHash === root) return true
  }

  // If no disclosed key produced a valid path, return false
  return disclosedKeys.length === 0 ? proof.length === 0 && root === sha256('') : false
}

module.exports = { merkleize, selectiveDisclose, verifyProof, sha256 }

File: sdk/go/merkle.go
package foodblock

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
)

// MerkleResult holds the result of merkle-izing a state object.
type MerkleResult struct {
	Root   string            `json:"root"`
	Leaves map[string]string `json:"leaves"`
	Tree   [][]string        `json:"tree"`
}

// DisclosureResult holds a selective disclosure with Merkle proof.
type DisclosureResult struct {
	Disclosed map[string]interface{} `json:"disclosed"`
	Proof     []ProofEntry           `json:"proof"`
	Root      string                 `json:"root"`
}

// ProofEntry is a sibling hash in a Merkle proof.
type ProofEntry struct {
	Hash     string `json:"hash"`
	Position string `json:"position"`
	Layer    int    `json:"layer"`
}

// Sha256Hex computes the SHA-256 hash of a string and returns it as hex.
func Sha256Hex(data string) string {
	h := sha256.Sum256([]byte(data))
	return hex.EncodeToString(h[:])
}

func canonicalMerkleValue(value interface{}) string {
	if value == nil {
		return "null"
	}
	switch v := value.(type) {
	case map[string]interface{}:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		b, _ := json.Marshal(v)
		return string(b)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// Merkleize creates a Merkle tree from a state object.
func Merkleize(state map[string]interface{}) MerkleResult {
	keys := make([]string, 0, len(state))
	for k := range state {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	leaves := make(map[string]string)
	for _, key := range keys {
		leaves[key] = Sha256Hex(key + ":" + canonicalMerkleValue(state[key]))
	}

	layer0 := make([]string, len(keys))
	for i, k := range keys {
		layer0[i] = leaves[k]
	}

	tree := [][]string{layer0}
	currentLayer := layer0

	for len(currentLayer) > 1 {
		var nextLayer []string
		for i := 0; i < len(currentLayer); i += 2 {
			if i+1 < len(currentLayer) {
				pair := []string{currentLayer[i], currentLayer[i+1]}
				sort.Strings(pair)
				nextLayer = append(nextLayer, Sha256Hex(pair[0]+pair[1]))
			} else {
				nextLayer = append(nextLayer, currentLayer[i])
			}
		}
		tree = append(tree, nextLayer)
		currentLayer = nextLayer
	}

	root := ""
	if len(currentLayer) > 0 {
		root = currentLayer[0]
	} else {
		root = Sha256Hex("")
	}

	return MerkleResult{Root: root, Leaves: leaves, Tree: tree}
}

// SelectiveDisclose creates a selective disclosure of specific fields with a Merkle proof.
func SelectiveDisclose(state map[string]interface{}, fieldNames []string) DisclosureResult {
	result := Merkleize(state)

	disclosed := make(map[string]interface{})
	for _, name := range fieldNames {
		if val, ok := state[name]; ok {
			disclosed[name] = val
		}
	}

	sortedKeys := make([]string, 0, len(state))
	for k := range state {
		sortedKeys = append(sortedKeys, k)
	}
	sort.Strings(sortedKeys)

	var proof []ProofEntry
	for _, name := range fieldNames {
		idx := -1
		for i, k := range sortedKeys {
			if k == name {
				idx = i
				break
			}
		}
		if idx == -1 {
			continue
		}

		currentIdx := idx
		for layer := 0; layer < len(result.Tree)-1; layer++ {
			layerNodes := result.Tree[layer]
			var siblingIdx int
			var position string
			if currentIdx%2 == 0 {
				siblingIdx = currentIdx + 1
				position = "right"
			} else {
				siblingIdx = currentIdx - 1
				position = "left"
			}
			if siblingIdx >= 0 && siblingIdx < len(layerNodes) {
				proof = append(proof, ProofEntry{
					Hash:     layerNodes[siblingIdx],
					Position: position,
					Layer:    layer,
				})
			}
			currentIdx = currentIdx / 2
		}
	}

	return DisclosureResult{Disclosed: disclosed, Proof: proof, Root: result.Root}
}

// VerifyProof verifies that disclosed fields and proof reconstruct the given Merkle root.
func VerifyProof(disclosed map[string]interface{}, proof []ProofEntry, root string) bool {
	if disclosed == nil || root == "" {
		return false
	}

	disclosedKeys := make([]string, 0, len(disclosed))
	for k := range disclosed {
		disclosedKeys = append(disclosedKeys, k)
	}
	sort.Strings(disclosedKeys)

	for _, key := range disclosedKeys {
		currentHash := Sha256Hex(key + ":" + canonicalMerkleValue(disclosed[key]))

		maxLayer := -1
		for _, p := range proof {
			if p.Layer > maxLayer {
				maxLayer = p.Layer
			}
		}

		byLayer := make(map[int][]ProofEntry)
		for _, p := range proof {
			byLayer[p.Layer] = append(byLayer[p.Layer], p)
		}

		for layer := 0; layer <= maxLayer; layer++ {
			entries := byLayer[layer]
			if len(entries) == 0 {
				continue
			}
			entry := entries[0]
			byLayer[layer] = entries[1:]

			var pair []string
			if entry.Position == "right" {
				pair = []string{currentHash, entry.Hash}
			} else {
				pair = []string{entry.Hash, currentHash}
			}
			sort.Strings(pair)
			currentHash = Sha256Hex(pair[0] + pair[1])
		}

		if currentHash == root {
			return true
		}
	}

	if len(disclosedKeys) == 0 {
		return len(proof) == 0 && root == Sha256Hex("")
	}
	return false
}
